[TASK] Use dependency injection in commands

brotkrueml opened the pull request:
- Autowiring/autoconfigure must be enabled so automatically dependency injection is possible
- The ClearIndexCommand now gets the IndexRepository injected to decouple database dependencies
  from the command
- The ClearIndexCommand/RemoveLockCommand gets rid of dependency to IndexerRunner only for logging.
  Now the logger is injected automatically.
- The shortcut logging methods are used for better readability.
- Unit tests are added for the commands.
- The IndexRepository was enhanced with methods for clearing the index.
- The IndexRepository gets the connection for the index table via constructor injection.
- Use str_contains instead of strstr for better readability. As TYPO3 v10 and v11 ship the PHP8.0
  symfony polyfill (also for legacy installations) this function can be used. The dependency was
  also added to composer.json for clarity.
- A GitHub workflow was added which lints the PHP files and runs the unit tests. I assume, the
  workflow actions must be enabled for the repository.
  The result of the workflow run is green, see:
  https://github.com/brotkrueml/ke_search/actions/runs/1859693267

Unit testing is now much easier and we get rid of GeneralUtility::makeInstance().

Unit tests can be executed via `.Build/bin/phpunit -c Tests/phpunit.xml.dist` or with
`composer test:unit`.

christianbltr commented:
Thanks for modernizing the commands!

I found the following issues:

(1) When the cache is flushed, the index is empty and the incremental indexer is started on the command line, an error is thrown:

```
vendor/bin/typo3cms cache:flush
vendor/bin/typo3cms ke_search:clearindex
vendor/bin/typo3cms ke_search:indexing --indexingMode=incremental
````

```
  [ ArgumentCountError ]                                                                                                                                                                                                   
  Too few arguments to function Tpwd\KeSearch\Domain\Repository\IndexRepository::__construct(), 0 passed in /var/www/html/public/typo3/sysext/core/Classes/Utility/GeneralUtility.php on line 3199 and exactly 1 expected  
 ```

(2) The same error is thrown in the ke_search backend module.

```
Too few arguments to function Tpwd\KeSearch\Domain\Repository\IndexRepository::__construct(), 0 passed in /var/www/html/public/typo3/sysext/core/Classes/Utility/GeneralUtility.php on line 3199 and exactly 1 expected
```

(3) The search box frontend plugin throws the error below.

```
PHP Warning: Undefined array key "prePostTagChar" in /var/www/html/public/typo3conf/ext/ke_search/Classes/Lib/Db.php line 323
```

This seems not to be only a "PHP 8 undefined array key" thing, because that error wasn't thrown before the patch. Maybe it's related to the addition of the type hint in the constructor of `Classes/Lib/Db.php`?

(4) The tests throw some deprecation messages when run under PHP 8.

```Remove Lock Command (Tpwd\KeSearch\Tests\Unit\Command\RemoveLockCommand)
 ✔ Start time of indexer is not defined then note is displayed
 ✔ Start time of indexer is defined then success message is displayed
 ✔ Error message is displayed when exception occurs
PHP Deprecated:  str_contains(): Passing null to parameter #1 ($haystack) of type string is deprecated in /var/www/html/public/typo3conf/ext/ke_search/Classes/Command/StartIndexerCommand.php on line 95
PHP Deprecated:  explode(): Passing null to parameter #2 ($string) of type string is deprecated in /var/www/html/public/typo3conf/ext/ke_search/Classes/Command/StartIndexerCommand.php on line 109

Deprecated: str_contains(): Passing null to parameter #1 ($haystack) of type string is deprecated in /var/www/html/public/typo3conf/ext/ke_search/Classes/Command/StartIndexerCommand.php on line 95

Deprecated: explode(): Passing null to parameter #2 ($string) of type string is deprecated in /var/www/html/public/typo3conf/ext/ke_search/Classes/Command/StartIndexerCommand.php on line 109
PHP Deprecated:  str_contains(): Passing null to parameter #1 ($haystack) of type string is deprecated in /var/www/html/public/typo3conf/ext/ke_search/Classes/Command/StartIndexerCommand.php on line 95
PHP Deprecated:  explode(): Passing null to parameter #2 ($string) of type string is deprecated in /var/www/html/public/typo3conf/ext/ke_search/Classes/Command/StartIndexerCommand.php on line 109

Deprecated: str_contains(): Passing null to parameter #1 ($haystack) of type string is deprecated in /var/www/html/public/typo3conf/ext/ke_search/Classes/Command/StartIndexerCommand.php on line 95

Deprecated: explode(): Passing null to parameter #2 ($string) of type string is deprecated in /var/www/html/public/typo3conf/ext/ke_search/Classes/Command/StartIndexerCommand.php on line 109
PHP Deprecated:  str_contains(): Passing null to parameter #1 ($haystack) of type string is deprecated in /var/www/html/public/typo3conf/ext/ke_search/Classes/Command/StartIndexerCommand.php on line 95
PHP Deprecated:  explode(): Passing null to parameter #2 ($string) of type string is deprecated in /var/www/html/public/typo3conf/ext/ke_search/Classes/Command/StartIndexerCommand.php on line 109

Deprecated: str_contains(): Passing null to parameter #1 ($haystack) of type string is deprecated in /var/www/html/public/typo3conf/ext/ke_search/Classes/Command/StartIndexerCommand.php on line 95

Deprecated: explode(): Passing null to parameter #2 ($string) of type string is deprecated in /var/www/html/public/typo3conf/ext/ke_search/Classes/Command/StartIndexerCommand.php on line 109
PHP Deprecated:  str_contains(): Passing null to parameter #1 ($haystack) of type string is deprecated in /var/www/html/public/typo3conf/ext/ke_search/Classes/Command/StartIndexerCommand.php on line 95
PHP Deprecated:  explode(): Passing null to parameter #2 ($string) of type string is deprecated in /var/www/html/public/typo3conf/ext/ke_search/Classes/Command/StartIndexerCommand.php on line 109

Deprecated: str_contains(): Passing null to parameter #1 ($haystack) of type string is deprecated in /var/www/html/public/typo3conf/ext/ke_search/Classes/Command/StartIndexerCommand.php on line 95

Deprecated: explode(): Passing null to parameter #2 ($string) of type string is deprecated in /var/www/html/public/typo3conf/ext/ke_search/Classes/Command/StartIndexerCommand.php on line 109
PHP Deprecated:  str_contains(): Passing null to parameter #1 ($haystack) of type string is deprecated in /var/www/html/public/typo3conf/ext/ke_search/Classes/Command/StartIndexerCommand.php on line 95
PHP Deprecated:  explode(): Passing null to parameter #2 ($string) of type string is deprecated in /var/www/html/public/typo3conf/ext/ke_search/Classes/Command/StartIndexerCommand.php on line 109

Deprecated: str_contains(): Passing null to parameter #1 ($haystack) of type string is deprecated in /var/www/html/public/typo3conf/ext/ke_search/Classes/Command/StartIndexerCommand.php on line 95

Deprecated: explode(): Passing null to parameter #2 ($string) of type string is deprecated in /var/www/html/public/typo3conf/ext/ke_search/Classes/Command/StartIndexerCommand.php on line 109
PHP Deprecated:  str_contains(): Passing null to parameter #1 ($haystack) of type string is deprecated in /var/www/html/public/typo3conf/ext/ke_search/Classes/Command/StartIndexerCommand.php on line 95
PHP Deprecated:  explode(): Passing null to parameter #2 ($string) of type string is deprecated in /var/www/html/public/typo3conf/ext/ke_search/Classes/Command/StartIndexerCommand.php on line 109
```

brotkrueml commented:
Sorry for the inconvenience.

> (1) When the cache is flushed, the index is empty and the incremental indexer is started on the command line, an error is thrown:
>   [ ArgumentCountError ]                                                                                                                                                                                                   
>   Too few arguments to function Tpwd\KeSearch\Domain\Repository\IndexRepository::__construct(), 0 passed in /var/www/html/public/typo3/sysext/core/Classes/Utility/GeneralUtility.php on line 3199 and exactly 1 expected  

This is fixed. For usage in GU::makeInstance() the class must be set to public.

> (2) The same error is thrown in the ke_search backend module.
> Too few arguments to function Tpwd\KeSearch\Domain\Repository\IndexRepository::__construct(), 0 passed in /var/www/html/public/typo3/sysext/core/Classes/Utility/GeneralUtility.php on line 3199 and exactly 1 expected

Same as (1).

> (3) The search box frontend plugin throws the error below.
> This seems not to be only a "PHP 8 undefined array key" thing, because that error wasn't thrown before the patch. Maybe it's related to the addition of the type hint in the constructor of `Classes/Lib/Db.php`?

Yup, the type hint in the annotation, which was previously given, was wrong. It is now corrected. But: the Filters class use GU::makeInstance() for instantiation of the Db class without giving the pObj as argument, so GU tries to inject it automatically, which fails of course. Workaround: set autoconfigure/autowire to false. The instantiation without pObj seems to be a mistake, but as this is not in the scope of the PR it should be handled in a separate change.

> (4) The tests throw some deprecation messages when run under PHP 8.

Good catch. Adjusted the configuration of phpunit, so that deprecations always throw an error, regardless of the php.ini setting of the underlying local system.

Checked it against my installation, seems to work. Please also have a look again.

If you merge this PR please squash it, so the timeline is not cluttered with the different WIP states and bug fixes.

christianbltr commented:
Thanks for the fix. I tested on my environment and everything worked fine. For the obviously wrong instantiation of the Db class in the Filter class I opened https://github.com/tpwd/ke_search/issues/78
